juju/model.py

```diff
@@ -43,32 +44,65 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
+    def _is_local_app(self, app_name):
+        return app_name == self.unit.app.name
+
     def get_unit(self, unit_name):
-        return self._cache.get(Unit, unit_name)
+        return self._cache.get(CacheKey(Unit, unit_name), unit_name, self._is_local_unit(unit_name), self._backend, self._cache)
+
+    def get_app(self, app_name):
+        return self._cache.get(CacheKey(Application, app_name), app_name, self._is_local_app(app_name), self._backend)
+
+
+class CacheKey:
+    def __init__(self, entity_type, *key_parts):
+        self.entity_type = entity_type
+        self.key_parts = key_parts
+
+    def __hash__(self):
+        return hash((self.entity_type, *self.key_parts))
```

Comment:
Why do we need the changes above?  Seems to be making things both more complex and less readable unnecessarily.